euphonic/readers/castep.py

```diff
@@ -53,15 +53,11 @@ def read_phonon_dos_data(
         weights = np.empty((0,))
         freqs = np.empty((0, n_branches))
         mode_grads = np.empty((0, n_branches))
-        while True:
-            frequency_block = _read_frequency_block(
+        while (frequency_block := _read_frequency_block(
```

Comment:
Does `_read_frequency_block` ever return falsy values which are not `None`? Should `_read_frequency_block` have a generator version for this?

Reply:
I don't think so, it usually returns a NamedTuple.

A generator would be nice. Something like:
```python
def _read_frequency_blocks(*args, **kwargs) -> Iterator[_FrequencyBlock]:
    while True:
        frequency_block = _read_frequency_block(*args, **kwargs)
    if frequency_block is None:
        raise StopIteration
    else:
        yield frequency_block
```

We could simplify further by putting the StopIteration into _read_frequency_block so it can only return one type.

Reply:
I think for debugging and testing purposes it is quite nice to have a "read one block" function available so would be reluctant to put the loop in there as well.

Reply:
It's nice to have a read one function for sure, what you wrote above (indented correctly, and possibly just with `break` rather than explicitly raising) would mean your loop just become `for blk in _read_frequency_block(...)` which is much clearer than the walrus.

Reply:
Done!